File: packages/uns-py/src/uns_kit/datahub_client.py
from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass
from typing import Any, Mapping, Optional

from .auth_client import AuthClient


@dataclass(frozen=True)
class LastValueResult:
    topic: str
    value: Any
    values: Optional[dict[str, Any]]
    uom: Optional[str]
    timestamp: Optional[str]
    data_group: Optional[str]
    age_ms: Optional[float]
    source: str

    @staticmethod
    def from_mapping(value: Mapping[str, Any]) -> "LastValueResult":
        raw_values = value.get("values")
        return LastValueResult(
            topic=str(value.get("topic") or ""),
            value=value.get("value"),
            values=dict(raw_values) if isinstance(raw_values, Mapping) else None,
            uom=value.get("uom") if isinstance(value.get("uom"), str) else None,
            timestamp=value.get("timestamp") if isinstance(value.get("timestamp"), str) else None,
            data_group=value.get("dataGroup") if isinstance(value.get("dataGroup"), str) else None,
            age_ms=float(value["ageMs"]) if isinstance(value.get("ageMs"), (int, float)) else None,
            source=str(value.get("source") or "miss"),
        )

    @property
    def hit(self) -> bool:
        return self.source == "cache"

    def to_dict(self) -> dict[str, Any]:
        return {
            "topic": self.topic,
            "value": self.value,
            "values": self.values,
            "uom": self.uom,
            "timestamp": self.timestamp,
            "dataGroup": self.data_group,
            "ageMs": self.age_ms,
            "source": self.source,
        }

class LastValueClientError(RuntimeError):
    def __init__(self, message: str, *, status_code: Optional[int] = None) -> None:
        super().__init__(message)
        self.status_code = status_code


class UnsClient:
    def __init__(
        self,
        base_url: str,
        *,
        api_base_path: str = "/api",
        token: Optional[str] = None,
        auth_client: Optional[AuthClient] = None,
        timeout: float = 10.0,
    ) -> None:
        self.api_base_path = self._normalize_base_path(api_base_path)
        stripped_base_url = base_url.rstrip("/")
        if self.api_base_path and stripped_base_url.endswith(self.api_base_path):
            self.api_url = stripped_base_url
            self.base_url = stripped_base_url[: -len(self.api_base_path)].rstrip("/")
        else:
            self.base_url = stripped_base_url
            self.api_url = f"{stripped_base_url}{self.api_base_path}"
        self.access_token = token
        self.timeout = timeout
        self.auth_client = auth_client
        self._opener = urllib.request.build_opener()

    def set_token(self, token: Optional[str]) -> None:
        self.access_token = token

    def ensure_token(self) -> None:
        if self.access_token:
            return
        try:
            if self.auth_client is None:
                self.auth_client = AuthClient.create()
            self.access_token = self.auth_client.get_access_token()

        except Exception as exc:
            raise RuntimeError("No access token available, please login again.") from exc

    def get(self, endpoint: str, params: Optional[dict[str, Any]] = None, *, base_url: Optional[str] = None, authorize: bool = True) -> dict[str, Any]:
        if authorize:
            self.ensure_token()
        query = urllib.parse.urlencode(params or {})
        suffix = f"?{query}" if query else ""
        return self._request_json(
            "GET",
            f"{self._build_url(endpoint, base_url=base_url)}{suffix}",
            token=self.access_token if authorize else None,
        )

    def post(self, endpoint: str, data: Optional[dict[str, Any]] = None, *, base_url: Optional[str] = None, authorize: bool = True) -> dict[str, Any]:
        if authorize:
            self.ensure_token()
        return self._request_json(
            "POST",
            self._build_url(endpoint, base_url=base_url),
            body=data or {},
            token=self.access_token if authorize else None,
        )

    def last_value(self, topics: str | list[str], *, token: Optional[str] = None) -> Optional[dict[str, dict[str, Any]]]:
        topic_list = [topics] if isinstance(topics, str) else topics
        if not topic_list:
            raise ValueError("topics must contain at least one topic.")
        if len(topic_list) > 500:
            raise ValueError("Maximum 500 topics per request.")
        effective_token = token
        if effective_token is None:
            self.ensure_token()
            effective_token = self.access_token
        try:
            payload = self._request_json(
                "POST",
                self._build_url("/batch/last"),
                body={"topics": topic_list},
                token=effective_token,
            )
        except LastValueClientError as exc:
            if exc.status_code == 404:
                return None
            raise
        raw_results = payload.get("results")
        if not isinstance(raw_results, list):
            raise LastValueClientError("Last-value response did not include a results array.")
        results = [
            LastValueResult.from_mapping(item)
            for item in raw_results
            if isinstance(item, Mapping)
        ]
        return {result.topic: result.to_dict() for result in results}

    def _request_json(
        self,
        method: str,
        url: str,
        *,
        body: Optional[dict[str, Any]] = None,
        token: Optional[str] = None,
    ) -> dict[str, Any]:
        data = None if body is None else json.dumps(body).encode("utf-8")
        request = urllib.request.Request(
            url,
            data=data,
            method=method,
            headers={
                "Accept": "application/json",
                **({"Content-Type": "application/json"} if data is not None else {}),
                **({"Authorization": f"Bearer {token}"} if token else {}),
            },
        )
        try:
            with self._opener.open(request, timeout=self.timeout) as response:
                decoded = response.read().decode("utf-8")
        except urllib.error.HTTPError as exc:
            message = exc.read().decode("utf-8", errors="replace")
            raise LastValueClientError(
                f"UNS request failed with HTTP {exc.code}: {message}",
                status_code=exc.code,
            ) from exc
        except urllib.error.URLError as exc:
            raise LastValueClientError(f"UNS request failed: {exc.reason}") from exc

        if not decoded:
            return {}
        try:
            parsed = json.loads(decoded)
        except json.JSONDecodeError as exc:
            raise LastValueClientError("UNS response was not valid JSON.") from exc
        if not isinstance(parsed, dict):
            raise LastValueClientError("UNS response must be a JSON object.")
        return parsed

    def _build_url(self, endpoint: str, *, base_url: Optional[str] = None) -> str:
        if endpoint.startswith("http://") or endpoint.startswith("https://"):
            return endpoint
        root = (base_url or self.api_url).rstrip("/")
        path = endpoint if endpoint.startswith("/") else f"/{endpoint}"
        if base_url is None and self.api_base_path and (
            path == self.api_base_path or path.startswith(f"{self.api_base_path}/")
        ):
            path = path[len(self.api_base_path):] or "/"
        return f"{root}{path}"

    @staticmethod
    def _normalize_base_path(value: str) -> str:
        stripped = value.strip()
        if not stripped:
            return ""
        return stripped if stripped.startswith("/") else f"/{stripped}"

File: packages/uns-core/src/tools/datahub/datahub-client.ts
import { AuthClient } from "../auth/auth-client.js";

export type LastValuePayload = {
  topic: string;
  value: unknown;
  values?: Record<string, unknown> | null;
  uom?: string | null;
  timestamp?: string | null;
  dataGroup?: string | null;
  ageMs?: number | null;
  source: string;
};

export class LastValueResult {
  readonly topic: string;
  readonly value: unknown;
  readonly values?: Record<string, unknown> | null;
  readonly uom?: string | null;
  readonly timestamp?: string | null;
  readonly dataGroup?: string | null;
  readonly ageMs?: number | null;
  readonly source: string;

  constructor(payload: LastValuePayload) {
    this.topic = payload.topic;
    this.value = payload.value;
    this.values = payload.values ?? null;
    this.uom = payload.uom ?? null;
    this.timestamp = payload.timestamp ?? null;
    this.dataGroup = payload.dataGroup ?? null;
    this.ageMs = payload.ageMs ?? null;
    this.source = payload.source;
  }

  static fromMapping(value: Record<string, unknown>): LastValueResult {
    const rawValues = value.values;
    return new LastValueResult({
      topic: typeof value.topic === "string" ? value.topic : "",
      value: value.value,
      values: rawValues && typeof rawValues === "object" && !Array.isArray(rawValues) ? (rawValues as Record<string, unknown>) : null,
      uom: typeof value.uom === "string" ? value.uom : null,
      timestamp: typeof value.timestamp === "string" ? value.timestamp : null,
      dataGroup: typeof value.dataGroup === "string" ? value.dataGroup : null,
      ageMs: typeof value.ageMs === "number" ? value.ageMs : null,
      source: typeof value.source === "string" ? value.source : "miss",
    });
  }

  get hit(): boolean {
    return this.source === "cache";
  }

  toObject(): LastValuePayload {
    return {
      topic: this.topic,
      value: this.value,
      values: this.values ?? null,
      uom: this.uom ?? null,
      timestamp: this.timestamp ?? null,
      dataGroup: this.dataGroup ?? null,
      ageMs: this.ageMs ?? null,
      source: this.source,
    };
  }
}

export class LastValueClientError extends Error {
  readonly statusCode?: number;

  constructor(message: string, statusCode?: number) {
    super(message);
    this.statusCode = statusCode;
  }
}

export type UnsClientOptions = {
  apiBasePath?: string;
  token?: string;
  timeoutMs?: number;
  authClient?: AuthClient;
};

export class UnsClient {
  private readonly apiBasePath: string;
  private readonly baseUrl: string;
  private readonly apiUrl: string;
  private readonly timeoutMs: number;
  private readonly authClient?: AuthClient;
  private accessToken?: string;

  constructor(baseUrl: string, options: UnsClientOptions = {}) {
    const apiBasePath = UnsClient.normalizeBasePath(options.apiBasePath ?? "/api");
    const strippedBase = baseUrl.replace(/\/$/, "");
    if (apiBasePath && strippedBase.endsWith(apiBasePath)) {
      this.apiUrl = strippedBase;
      this.baseUrl = strippedBase.slice(0, -apiBasePath.length).replace(/\/$/, "");
    } else {
      this.baseUrl = strippedBase;
      this.apiUrl = `${strippedBase}${apiBasePath}`;
    }
    this.apiBasePath = apiBasePath;
    this.timeoutMs = options.timeoutMs ?? 10_000;
    this.authClient = options.authClient;
    this.accessToken = options.token;
  }

  setToken(token?: string): void {
    this.accessToken = token;
  }

  async ensureToken(): Promise<string | undefined> {
    if (this.accessToken) return this.accessToken;
    if (!this.authClient) return undefined;
    const token = await this.authClient.getAccessToken();
    this.accessToken = token;
    return token;
  }

  async get(endpoint: string, params?: Record<string, string | number | boolean>, options: { baseUrl?: string; authorize?: boolean } = {}): Promise<Record<string, unknown>> {
    const authorize = options.authorize ?? true;
    const token = authorize ? await this.ensureToken() : undefined;
    const search = params ? `?${new URLSearchParams(this.stringifyQueryParams(params))}` : "";
    return this.requestJson("GET", `${this.buildUrl(endpoint, options.baseUrl)}${search}`, undefined, token);
  }

  async post(endpoint: string, body?: Record<string, unknown>, options: { baseUrl?: string; authorize?: boolean } = {}): Promise<Record<string, unknown>> {
    const authorize = options.authorize ?? true;
    const token = authorize ? await this.ensureToken() : undefined;
    return this.requestJson("POST", this.buildUrl(endpoint, options.baseUrl), body ?? {}, token);
  }

  async lastValue(topics: string | string[], options: { token?: string } = {}): Promise<Record<string, LastValuePayload> | null> {
    const topicList = Array.isArray(topics) ? topics : [topics];
    if (!topicList.length) {
      throw new Error("topics must contain at least one topic.");
    }
    if (topicList.length > 500) {
      throw new Error("Maximum 500 topics per request.");
    }
    const token = options.token ?? (await this.ensureToken());
    try {
      const payload = await this.requestJson("POST", this.buildUrl("/batch/last"), { topics: topicList }, token);
      const rawResults = payload.results;
      if (!Array.isArray(rawResults)) {
        throw new LastValueClientError("Last-value response did not include a results array.");
      }
      const results = rawResults
        .filter((item): item is Record<string, unknown> => !!item && typeof item === "object" && !Array.isArray(item))
        .map((item) => LastValueResult.fromMapping(item));
      return Object.fromEntries(results.map((result) => [result.topic, result.toObject()]));
    } catch (error) {
      if (error instanceof LastValueClientError && error.statusCode === 404) {
        return null;
      }
      throw error;
    }
  }

  private async requestJson(
    method: "GET" | "POST",
    url: string,
    body?: Record<string, unknown>,
    token?: string,
  ): Promise<Record<string, unknown>> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const resp = await fetch(url, {
        method,
        headers: {
          "Accept": "application/json",
          ...(body ? { "Content-Type": "application/json" } : {}),
          ...(token ? { "Authorization": `Bearer ${token}` } : {}),
        },
        ...(body ? { body: JSON.stringify(body) } : {}),
        signal: controller.signal,
      });
      const text = await resp.text();
      if (!resp.ok) {
        throw new LastValueClientError(
          `UNS request failed with HTTP ${resp.status}: ${text || resp.statusText}`,
          resp.status,
        );
      }
      if (!text) return {};
      const parsed = JSON.parse(text);
      if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
        throw new LastValueClientError("UNS response must be a JSON object.");
      }
      return parsed as Record<string, unknown>;
    } catch (error: any) {
      if (error instanceof LastValueClientError) {
        throw error;
      }
      if (error?.name === "AbortError") {
        throw new LastValueClientError("UNS request timed out.");
      }
      throw new LastValueClientError(`UNS request failed: ${error?.message ?? String(error)}`);
    } finally {
      clearTimeout(timeout);
    }
  }

  private buildUrl(endpoint: string, baseUrl?: string): string {
    if (endpoint.startsWith("http://") || endpoint.startsWith("https://")) {
      return endpoint;
    }
    const root = (baseUrl ?? this.apiUrl).replace(/\/$/, "");
    let path = endpoint.startsWith("/") ? endpoint : `/${endpoint}`;
    if (!baseUrl && this.apiBasePath && (path === this.apiBasePath || path.startsWith(`${this.apiBasePath}/`))) {
      path = path.slice(this.apiBasePath.length) || "/";
    }
    return `${root}${path}`;
  }

  private stringifyQueryParams(params: Record<string, string | number | boolean>): Record<string, string> {
    return Object.fromEntries(Object.entries(params).map(([key, value]) => [key, String(value)]));
  }

  private static normalizeBasePath(value: string): string {
    const stripped = value.trim();
    if (!stripped) return "";
    return stripped.startsWith("/") ? stripped : `/${stripped}`;
  }
}
